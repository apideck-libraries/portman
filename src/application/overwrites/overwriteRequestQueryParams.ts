import { PostmanMappedOperation } from 'postman'
import { OverwriteQueryParamConfig } from 'types'

/**
 * Overwrite Postman request query params with values defined by the portman testsuite
 * @param overwriteValues
 * @param pmOperation
 */
export const overwriteRequestQueryParams = (
  overwriteValues: OverwriteQueryParamConfig[],
  pmOperation: PostmanMappedOperation
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
): PostmanMappedOperation => {
  // Early exit if overwrite values are not defined
  if (!(overwriteValues instanceof Array)) return pmOperation

  // Early exit if request url query is not defined
  if (!pmOperation.item?.request?.url?.query) return pmOperation

  pmOperation.item.request.url.query.each(pmQueryParam => {
    // Overwrite values for Keys
    overwriteValues.forEach(overwriteValue => {
      // Skip keys when no overwrite is defined
      if (!(overwriteValue.key && pmQueryParam.key && overwriteValue.key === pmQueryParam.key)) {
        return
      }

      // Test suite - Overwrite/extend query param value
      if (overwriteValue?.value && pmQueryParam?.value) {
        const orginalValue = pmQueryParam.value
        let newValue = overwriteValue.value

        if (overwriteValue.overwrite === false) {
          newValue = orginalValue + newValue
        }
        pmQueryParam.value = newValue
      }

      // Test suite - Disable query param
      if (overwriteValue?.disable === true) {
        pmQueryParam.disabled = true
      }

      // Set Postman query param
      pmOperation.item.request.url.query.upsert(pmQueryParam)
    })
  })
  // Test suite - Remove query param
  overwriteValues
    .filter(({ remove }) => remove)
    .map(paramToRemove => {
      pmOperation.item.request.url.removeQueryParams(paramToRemove.key)
    })

  return pmOperation
}
