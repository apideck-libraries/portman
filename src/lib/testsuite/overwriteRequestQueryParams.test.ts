import { getPostmanMappedOperation } from '../../../__tests__/getPostmanMappedOperation'
import { overwriteRequestQueryParams } from './overwriteRequestQueryParams'

describe('overwriteRequestQueryParams', () => {
  it('should overwrite the request query param', async () => {
    const overwriteValues = [
      {
        key: 'raw',
        value: 'foo'
      }
    ]
    const pmOperation = await getPostmanMappedOperation()
    const result = overwriteRequestQueryParams(overwriteValues, pmOperation)
    expect(result.item.request.url.query).toMatchSnapshot()
  })

  it('should disable the request query param', async () => {
    const overwriteValues = [
      {
        key: 'raw',
        disable: true
      }
    ]
    const pmOperation = await getPostmanMappedOperation()
    const result = overwriteRequestQueryParams(overwriteValues, pmOperation)
    expect(result.item.request.url.query).toMatchSnapshot()
  })

  it('should append the request path variable when overwrite is false', async () => {
    const overwriteValues = [
      {
        key: 'raw',
        value: 'foo',
        overwrite: false
      }
    ]
    const pmOperation = await getPostmanMappedOperation()
    const result = overwriteRequestQueryParams(overwriteValues, pmOperation)
    expect(result.item.request.url.query).toMatchSnapshot()
  })

  it('should return unaltered if key not found', async () => {
    const overwriteValues = [
      {
        key: 'not-a-variable',
        value: 'foo-bar-baz'
      }
    ]

    const pmOperation = await getPostmanMappedOperation()
    const result = overwriteRequestQueryParams(overwriteValues, pmOperation)

    expect(result.item.request.url.query).toEqual(pmOperation.item.request.url.query)
  })
})
