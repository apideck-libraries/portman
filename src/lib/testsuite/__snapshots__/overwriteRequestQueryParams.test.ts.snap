// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overwriteRequestQueryParams should append the request query param when overwrite is false 1`] = `
Array [
  Object {
    "description": Object {
      "content": "Include raw response. Mostly used for debugging purposes",
      "type": "text/plain",
    },
    "disabled": false,
    "key": "raw",
    "value": "truefoo",
  },
]
`;

exports[`overwriteRequestQueryParams should disable the request query param 1`] = `
Array [
  Object {
    "description": Object {
      "content": "Include raw response. Mostly used for debugging purposes",
      "type": "text/plain",
    },
    "disabled": true,
    "key": "raw",
    "value": "true",
  },
]
`;

exports[`overwriteRequestQueryParams should overwrite the request query param 1`] = `
Array [
  Object {
    "description": Object {
      "content": "Include raw response. Mostly used for debugging purposes",
      "type": "text/plain",
    },
    "disabled": false,
    "key": "raw",
    "value": "foo",
  },
]
`;

exports[`overwriteRequestQueryParams should remove to the query param when remove is true 1`] = `Array []`;
